src/main.rs

```diff
@@ -0,0 +1,91 @@
+use std::time::Duration;
+
+use clap::Parser;
+
+use nomad_external_dns::hetzner_dns;
+use reqwest::Url;
+use tokio::time::sleep;
+
+use nomad_external_dns::config::{Config, DnsProvider};
+use nomad_external_dns::consul::ConsulClient;
+use nomad_external_dns::dns_trait::DnsProviderTrait;
+
+#[tokio::main]
+async fn main() {
+    println!("Starting up Nomad External DNS");
+
+    let config = match Config::try_parse() {
+        Ok(config) => config,
+        Err(e) => {
+            eprintln!("Failed to parse configuration: {}", e);
+            return;
+        }
+    };
+    println!("Configuration parsed successfully");
+
+    // Initialize Consul Client
+    let consul_client = loop {
+        match ConsulClient::new(
+            Url::parse(&config.consul_address).expect("Invalid URL"),
+            config.consul_datacenter.clone(),
+        ) {
+            Ok(client) => break client,
+            Err(e) => {
+                eprintln!("Failed to create Consul client: {}. Retrying in 100ms", e);
+                sleep(Duration::from_millis(100)).await;
+            }
+        }
+    };
+    println!("Consul client created successfully");
+
+    // Acquire Lock
```

Comment:
Once we store the state properly and can delete old records we need to run what follows until the end of the function in a loop. Then we want to do [long-polling to wait for changes](https://developer.hashicorp.com/consul/api-docs/features/blocking) to the Consul services.  The way to do that is by having a state variable `consul_index` which starts out being set to `None`.  Whenever we fetch the list of services from Consul:

1. If `consul_index` is `Some` we set the query parameter `index` to its value.
2. If we got a successful API response we take the value from the `X-Consul-Index` header and assign it to `consul_index`.

Setting the `index` query parameter blocks the request until either a change to the filtered results is detected or the request times out.  If the request times out we simply start the loop from the top.